### ISSUE TITLE
FastAPI deployed with hypercorn in GCP Cloud Run returning 503 sporadically

### ISSUE THREAD

**Issue by bgregoinductiva** [pgjones/hypercorn#221]
I have a FastAPI project deployed in Cloud Run using the hypercorn server. I'm using Uvloop as the event loop and leaving the other configurations with default values:

hypercorn app.main:app --bind 0.0.0.0:80 --worker-class uvloop
 
Here are the Cloud Run configurations:
- Memory: 1 GiB
- CPU: 1
- Maximum concurrent requests per instance: 80
- CPU is only allocated during request processing
- Minimum number of instances: 1
- Maximum number of instances: 30
- Startup CPU boost
- Use HTTP/2 end-to-end
 
When I get a peak of concurrent requests during integration testing, about 30, I usually get a 503, and then a new instance is started.
 
Has anyone faced a similar problem before?

Thanks in advance.

**Comment by nabheet** [comment]
Yes, based on what I have learnt so far, your instance was terminated because it accessed more memory that its defined limit. 

Even though [this](https://cloud.google.com/run/docs/container-contract#forced-termination) says that Cloud Run will return a 500. In my testing I was able to prove the it actually returns a 503. Their documentation leaves a lot to be desired.

Hope this helps.

**Comment by nielsbox** [comment]
We have the same issue, only at 40% memory usage at 99%.

**Comment by nielsbox** [comment]
Update: we isolated the issue to only HTTP/2. HTTP/1 seems to be fine.

**Comment by pgjones** [comment]
> Update: we isolated the issue to only HTTP/2. HTTP/1 seems to be fine.

Is the HTTP/1 traffic encrypted? There seems to be an asyncio memory leak with SSL

**Comment by nielsbox** [comment]
> > Update: we isolated the issue to only HTTP/2. HTTP/1 seems to be fine.
> 
> Is the HTTP/1 traffic encrypted? There seems to be an asyncio memory leak with SSL

Cloudrun terminates TLS.
https://cloud.google.com/run/docs/container-contract#tls

**Comment by nabheet** [comment]
Also, I hate to admit this in public, but I wasn't closing SQL connections in the health check endpoint so that was leaking file descriptors. This was causing our Cloud Run containers to crash without log events returning a 503 from the Cloud Run LB.

So another thing to check would be your file descriptor count.

**Comment by rushilsrivastava** [comment]
> So another thing to check would be your file descriptor count.

Curious how you profiled this with Cloud Run

**Comment by nabheet** [comment]
> Curious how you profiled this with Cloud Run

create an endpoint that executes the following shell command in the container:
```
lsof | wc -l
```

Also, I think Opentelemetry has a metric for open file descriptor count also.

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — I couldn't get this one to surface on current `main` and the thread has gone quiet.

No isolated MRE attributing this to hypercorn; users found their own causes.

Evidence:
- [comment](https://github.com/pgjones/hypercorn/issues/221#issuecomment-2143812064) — nabheet self-reported root-cause was their own SQL-connection leak / file descriptors — not Hypercorn.
- [comment](https://github.com/pgjones/hypercorn/issues/221#issuecomment-2657859213) — Last activity Feb 2025; thread devolved into FD-count debugging.

Happy to be wrong if there's a current MRE I missed; otherwise would you mind closing this out and reopening with a fresh repro if it returns?

---
*Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually.*
